samples/bluetooth/hci_uart_async/src/hci_uart_async.c

```diff
@@ -0,0 +1,371 @@
+		if (err) {
+			net_buf_unref(buf);
+			return;
+		}
+		LOG_HEXDUMP_DBG(hdr_buf, hdr_size, "h2c: hci hdr");
+
+		/* Read HCI payload. */
+		payload_size = hci_payload_size(hdr_buf, h4_type);
+
+		LOG_DBG("h2c: payload_size %u", payload_size);
+
+		if (payload_size <= net_buf_tailroom(buf)) {
+			uint8_t *payload_dst = net_buf_add(buf, payload_size);
+
+			err = uart_h2c_rx(payload_dst, payload_size);
+			if (err) {
+				net_buf_unref(buf);
+				return;
+			}
+			LOG_HEXDUMP_DBG(payload_dst, payload_size, "h2c: hci payload");
+		} else {
+			uint8_t *discard_dst;
+			uint16_t discard_size;
+
+			/* Discard oversize packet. */
+			LOG_WRN("h2c: Discarding oversize h4_type %d payload_size %d.", h4_type,
+				payload_size);
+
+			net_buf_reset(buf);
+			discard_dst = net_buf_tail(buf);
+			discard_size = net_buf_max_len(buf);
+
+			while (payload_size) {
+				uint16_t read_size = MIN(payload_size, discard_size);
+
+				err = uart_h2c_rx(discard_dst, read_size);
+				if (err) {
+					net_buf_unref(buf);
+					return;
+				}
+
+				payload_size -= read_size;
+			}
+
+			net_buf_unref(buf);
+			buf = NULL;
+		}
+
+		LOG_DBG("h2c: packet done");
+
+		/* Route buf to Controller. */
+		if (buf) {
+			err = bt_send(buf);
+			if (err) {
+				/* This is not a transport error. */
+				LOG_ERR("bt_send err %d", err);
+				net_buf_unref(buf);
+				buf = NULL;
+			}
+		}
+
+		k_yield();
+	}
+}
+
+static void h2c_thread_entry(void *p1, void *p2, void *p3)
+{
+	k_thread_name_set(k_current_get(), "HCI TX (h2c)");
+
+	for (;;) {
+		LOG_DBG("Synchronized");
+		h2c_h4_transport();
+		LOG_WRN("Desynchronized");
+		send_hw_error();
+		recover_sync_by_reset_pattern();
+	}
+}
+
+void callback(const struct device *dev, struct uart_event *evt, void *user_data)
+{
+	ARG_UNUSED(user_data);
+
+	if (evt->type == UART_RX_DISABLED) {
+		k_poll_signal_raise(&uart_h2c_rx_sig, 0);
+	}
+	if (evt->type == UART_RX_STOPPED) {
+		k_poll_signal_raise(&uart_h2c_rx_sig, evt->data.rx_stop.reason);
+	}
+	if (evt->type == UART_TX_DONE) {
+		k_poll_signal_raise(&uart_c2h_tx_sig, 0);
+	}
+}
+
+static int hci_uart_init(void)
+{
+	int err;
+
+	k_poll_signal_init(&uart_h2c_rx_sig);
+	k_poll_signal_init(&uart_c2h_tx_sig);
+
+	LOG_DBG("");
+
+	if (!device_is_ready(hci_uart_dev)) {
+		LOG_ERR("HCI UART %s is not ready", hci_uart_dev->name);
+		return -EINVAL;
+	}
+
+	BUILD_ASSERT(IS_ENABLED(CONFIG_UART_ASYNC_API));
+	err = uart_callback_set(hci_uart_dev, callback, NULL);
+
+	/* Is CONFIG_UART_ASYNC_API not implemented/enabled for the
+	 * selected `hci_uart_dev`?
+	 */
+	__ASSERT(!err, "err %d", err);
+
+	return 0;
+}
+
+SYS_INIT(hci_uart_init, APPLICATION, CONFIG_KERNEL_INIT_PRIORITY_DEVICE);
+
+const struct {
+	uint8_t h4;
+	struct bt_hci_evt_hdr hdr;
+	struct bt_hci_evt_cmd_complete cc;
+} __packed cc_evt = {
+	.h4 = H4_EVT,
+	.hdr = {.evt = BT_HCI_EVT_CMD_COMPLETE, .len = sizeof(struct bt_hci_evt_cmd_complete)},
+	.cc = {.ncmd = 1, /* TODO: Should this value come from a config? */
```

Comment:
would be good to get rid of `TODO`

Reply:
I removed the TODO. It was like this in the original `hci_uart` sample. I assume `CONFIG_BT_WAIT_NOP` wants it like this.